### DISCUSSION TITLE
How to create an uppercase only input ?

### DISCUSSION THREAD

**Discussion by Exifers** [react-hook-form/react-hook-form#7867] [Q&A] [answered]
Hi, thanks for the amazing lib.

Though, after looking at all the docs, I can't find a way to do a classic *transformed value* input, such as one that would set its content to uppercase on each change event.

Here is a typical implementation in vanilla React :  
```jsx
const [value, setValue] = useState('')

return (
  <input type="text" value={value} onChange={event => setValue(event.target.value.toUpperCase())}/>
)
```

Now to do this with RHF, I tried `setValueAs` but it applies the transform only on the submitted data while the input still has lowercase letters.
See sandbox here : https://codesandbox.io/s/react-hook-form-get-started-forked-vqmzde?file=/src/index.js 

I also tried : 
```jsx
<input
  {...register("city")}
  onChange={(event) => {
    event.target.value = event.target.value.toUpperCase();
  }}
/>
```
but this creates some cursor positioning problems.

To reproduce :   
- type characters in the first input
- the cursor is at the end of the input
- move the cursor at the beginning
- type a lower case character

**Observed result :**  
The cursor is teleported at the end of the input
**Expected result :**
The cursor is at the second position in the input

What is the recommended way to create such an input with RHF ?
Thanks !

**Comment by Moshyfawn** [comment]
I'd suggest to use CSS `text-transform: uppercase` to style the input's text and transform the actual input value to uppercase `onSubmit`

**Reply by Exifers** [reply]
Thanks @Moshyfawn ! is there a way for the generic use case of any kind of transform through a javascript function ?

It could be : 
- other types of spell cases not supported by CSS, such as sentence case
- prevent input of certain types of characters, for example this will prevent spaces : `event.target.value.replace(' ', '')`
- autoformat input on each keystroke, for example this removes leading spaces : `event.target.value.replace(/^\s+/g,'')`

I guess there are other cases too.

Where to put such a function to transform the value ?

Each time when I try with these functions I get the cursor positioning bug.

**Reply by Moshyfawn** [reply]
TL;DR To transform the input's value use controlled component _(`Controller` / `useController` in case of React Hook Form)_

I don't think you can achieve this functionality with **an uncontrolled** input, which is what your input's using the `register` method are. To manipulate the input's value, it needs to **be fed it's transformed value**, which means, it's gotta be **a controlled** input.

The **uncontrolled** input means that **the input's value is the source of truth**. **To manipulate and update** the input's value, **it has to be outside the input**, which is React state in case of React and React Hook Form.

Even in your example of a "typical implementation" you're using React state to pass the updated value back into your input.
```jsx
const [value, setValue] = useState('')

return (
  <input type="text" value={value} onChange={event => setValue(event.target.value.toUpperCase())}/>
)
```

Also, the cursor behavior when mutating the `event.target.value` is expected and it's not React Hook Form specific. It's the way native Web components work, so I'd advice against it.

**Reply by Exifers** [reply]
Ok thanks @Moshyfawn ! very clear.

**Reply by guizzo** [reply]
@Exifers @Moshyfawn  This is not a good solution in my opinion, because with this approach the user input is not really capitalized; it's just appearance until the data passes through the submitHandler; only after this does the user input really change to an uppercase string.
For example, if the input were subjected to a case sensitive RegExp that only accepted uppercase characters, this kind of approach would not allow the data to pass validation, even though the data is displayed in uppercase on the screen.

Don't you agree?

**Reply by Moshyfawn** [reply]
In case you need to transform user input, use `Controller` / `useController` [as mentioned in this thread](https://github.com/react-hook-form/react-hook-form/discussions/7867#discussioncomment-2212314)

**Reply by oPolotniuk** [reply]
@Exifers did it help you ? I'm having the same problem

**Comment by soulr344** [comment, marked as answer]
Ive found a simple solution for this. Basically store the current cursor position and forcefully set the cursor position when each value changes.

https://gist.github.com/soulr344/8869f597c86e30a1d4b8655aeab00da9

**Reply by Exifers** [reply]
Yes actually the cursor jumping happens with React alone (without RHF).
There was an issue on the react github repo talking about this, and the solution was to track the cursor position as you said.

**Reply by sheepiiHD** [reply]
There is absolutely no way this is an accepted answer lmao. The solution shouldn't be to simulate user actions, it should be to just manage the state better. You can transform the text to uppercase, and display/submit it via uppercase.

**Reply by soulr344** [reply]
@sheepiiHD your answer has some limitations, such as when you try to copy text from an input which has transform uppercase, the copied text won't be in uppercase, rather in the original case you typed in.

Another way to achieve this is with text transform uppercase and set the value to uppercase onBlur event, and handle the copy event accordingly. The solution above is great when you've bad compositions on an already finished projects. For new projects I'd stick to onBlur method I described earlier. If you have a better solution, do tell.
